src/backend/distributed/utils/query_utils.c

```diff
@@ -0,0 +1,133 @@
+/*-------------------------------------------------------------------------
+ *
+ * query_utils.c
+ *
+ * Query-walker utility functions to be used to construct a logical plan
+ * tree from the given query tree structure.
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ * $Id$
+ *
+ ****-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "nodes/nodeFuncs.h"
+#include "distributed/query_utils.h"
+#include "distributed/query_pushdown_planning.h"
+
+/* This function will be used to walk over a tree to gather entries.
+ * Execution is parameterized by passing walkerMode flag via ExtractParams
+ * as we cannot pass more than one parameter to query_tree_walker
+ */
+bool
+ExtractRangeTableList(Node *node, ExtractParams *params)
+{
+	/* get parameters from ExtractParams context */
+	List **rangeTableRelationList = params->rangeTableList;
+	ExtractMode walkerMode = params->walkerMode;
+
+	bool walkIsComplete = false;
+
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, RangeTblEntry) && walkerMode != EXT_INDEX_ENTRIES)
+	{
+		RangeTblEntry *rangeTable = (RangeTblEntry *) node;
+
+		if (walkerMode == EXT_ALL_ENTRIES ||
+			(rangeTable->rtekind == RTE_RELATION && rangeTable->relkind != RELKIND_VIEW))
+		{
+			(*rangeTableRelationList) = lappend(*rangeTableRelationList, rangeTable);
+		}
+	}
+	else if (IsA(node, RangeTblRef) && walkerMode == EXT_INDEX_ENTRIES)
+	{
+		int rangeTableIndex = ((RangeTblRef *) node)->rtindex;
+		(*rangeTableRelationList) = lappend_int(*rangeTableRelationList, rangeTableIndex);
+	}
+	else if (IsA(node, Query) && walkerMode != EXT_INDEX_ENTRIES)
+	{
+		Query *query = (Query *) node;
+
+		if (walkerMode == EXT_ALL_ENTRIES || query->hasSubLinks || query->cteList ||
```

Comment:
I don't think we should have the walkerMode comparison here. The if statement is an optimisation, because we know the query tree does not contain RTEs other than the ones in rangeTableList if these conditions hold. It would be nice to apply that optimisation in all cases.